Mathlib/Data/Vector/Snoc.lean

```diff
@@ -0,0 +1,167 @@
+/-
+Copyright (c) 2023 Alex Keizer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex Keizer
+-/
+import Mathlib.Data.Vector.Basic
+
+/-!
+  This file establishes a `snoc : Vector α n → α → Vector α (n+1)` operation, that appends a single
+  element to the back of a vector.
+
+  It provides a collection of lemmas that show how different `Vector` operations reduce when their
+  argument is `snoc xs x`.
+
+  Also, an alternative, reverse, induction principle is added, that breaks down a vector into
+  `snoc xs x` for its inductive case. Effectively doing induction from right-to-left
+-/
+
+namespace Vector
+
+  /-- Append a single element to the end of a vector -/
+  def snoc : Vector α n → α → Vector α (n+1) :=
+    fun xs x => append xs (x ::ᵥ Vector.nil)
+
+
+  /-!
+  ## Simplification lemmas
+  -/
+  section Simp
+    variable (xs : Vector α n)
+
+  @[simp]
+  theorem snoc_cons : (x ::ᵥ xs).snoc y = x ::ᵥ (xs.snoc y) :=
+    rfl
+
+  @[simp]
+  theorem snoc_nil : (nil.snoc x) = x ::ᵥ nil :=
+    rfl
+
+  @[simp]
+  theorem reverse_cons : reverse (x ::ᵥ xs) = (reverse xs).snoc x := by
+    cases xs
+    simp only [reverse, cons, toList_mk, List.reverse_cons, snoc]
+    congr
+
+  @[simp]
+  theorem reverse_snoc : reverse (xs.snoc x) = x ::ᵥ (reverse xs) := by
+    cases xs
+    simp only [reverse, snoc, cons, toList_mk]
+    congr
+    simp[toList, (·++·), Vector.append, Append.append]
+    rfl
+
+
+  theorem replicate_succ_to_snoc (val : α) :
+      replicate (n+1) val = (replicate n val).snoc val := by
+    clear xs
+    induction n
+    case zero => rfl
+    case succ n ih =>
+      rw[replicate_succ]
+      conv => {
+        rhs; rw[replicate_succ]
+      }
+      rw[snoc_cons, ih]
+
+  end Simp
+
+
+  /-!
+  ## Reverse induction principle
+  -/
+  section Induction
+
+  /-- Define `C v` by *reverse* induction on `v : Vector α n`.
+      That is, break the vector down starting from the right-most element, using `snoc`
+
+      This function has two arguments: `nil` handles the base case on `C nil`,
+      and `snoc` defines the inductive step using `∀ x : α, C xs → C (xs.snoc x)`.
+
+      This can be used as `induction v using Vector.revInductionOn`. -/
+  @[elab_as_elim]
+  def revInductionOn {C : ∀ {n : ℕ}, Vector α n → Sort _} {n : ℕ} (v : Vector α n)
+                      (nil : C nil)
+                      (snoc : ∀ {n : ℕ} (xs : Vector α n) (x : α), C xs → C (xs.snoc x))
+                      : C v :=
```

Comment:
This indentation style is not usual. We just indent the theorem statement with four spaces usually, with occasional extra two space indentations for readability.